### PR TITLE
fix(check-merge-safety): only set merge safety status to failure if it was previous success

### PR DESCRIPTION
**Pull request by aschwenn**
### :pencil: Description
GitHub has a limit of 1000 statuses that can be set per commit. Since `check-merge-safety` runs so many times on large-scale repos, it lately has ended up accidentally hitting this limit, which fails the rest of the job so it doesn't finish setting statuses. Here, I check for an existing failure status before attempting to set a failure status. We could also perform the same check for success statuses, but that's not really a big problem right now; the issue is stale branches that are having the failure status set a thousand times.

### :link: Related Issues
-

### CODE REVIEW

File: src/helpers/check-merge-safety.ts [hunk 1]
```diff
@@ -21,6 +21,7 @@ import { map } from 'bluebird';
 import { setCommitStatus } from './set-commit-status';
 import * as core from '@actions/core';
 
+const maxBranchNameLength = 50;
```
Comment (danadajian) [inline src/helpers/check-merge-safety.ts:24]: Just curious, did the linter get mad at defining this after usage?

File: src/helpers/check-merge-safety.ts [hunk 2]
```diff
@@ -61,6 +75,18 @@ const handlePushWorkflow = async (inputs: CheckMergeSafety) => {
...
+const checkForExistingFailureStatus = async (pullRequest: PullRequest, context: string) => {
+  const { data } = await octokit.repos.getCombinedStatusForRef({
+    ...githubContext.repo,
+    ref: pullRequest.head.sha
+  });
+  if (data.state === 'failure') {
+    const existingContext = data.statuses.find(status => status.context === context);
+    return Boolean(existingContext);
+  }
+  return false;
```
Comment (danadajian) [inline src/helpers/check-merge-safety.ts:87]: I see the micro-optimization here, but this seems simpler. Am I missing something?
```suggestion
  return Boolean(data.statuses.find(status => status.context === context));
```

File: src/helpers/check-merge-safety.ts (hunk 1 above)
Comment (aschwenn) [reply src/helpers/check-merge-safety.ts:24]: nah didn't see anything come up

File: src/helpers/check-merge-safety.ts (hunk 2 above)
Comment (aschwenn) [reply src/helpers/check-merge-safety.ts:87]: statuses will include `Merge Safety` even if the state is `success`, so we need to ensure we're looking at a `failure` before deciding not to re-set the status to `failure`. for a `failure`, failed statuses are listed, and for a `success`, all manually-set successful statuses are listed (afaik)

Comment (danadajian) [reply src/helpers/check-merge-safety.ts:87]: ah ok, makes sense!

File: src/helpers/check-merge-safety.ts [hunk 3]
```diff
@@ -44,13 +45,26 @@ export const checkMergeSafety = async (inputs: CheckMergeSafety) => {
...
   const { state, message } = await getMergeSafetyStateAndMessage(pullRequest, inputs);
-  await setCommitStatus({
-    sha: pullRequest.head.sha,
-    state,
-    context,
-    description: message,
-    ...githubContext.repo
-  });
+  const hasExistingFailureStatus = await checkForExistingFailureStatus(pullRequest, context);
+  if (hasExistingFailureStatus && state === 'failure') {
```
Comment (danadajian) [inline src/helpers/check-merge-safety.ts:49]: Isn't this `state === 'failure'` check already happening inside `checkForExistingFailureStatus`?

Comment (aschwenn) [reply src/helpers/check-merge-safety.ts:49]: `state` here is the current status, whereas the value from `checkForExistingFailureStatus` is coming from the previous status

basically added this check in to say that in the case where there is an existing `failure` status, only skip if the new status is also a `failure` (it's almost impossible for the status to flip from from `failure` to `success` unless someone were to force-push the target branch, just wanted to make sure our bases were covered)

Comment (danadajian) [reply src/helpers/check-merge-safety.ts:49]: Gotcha - I was getting confused between this state and the individual PR "combined" state

**Review by danadajian — APPROVED** [review]

**Pull request merged**